elab-stlc-letrec/Surface.ml

```diff
@@ -0,0 +1,268 @@
+  | FunType (ty1, ty2) ->
+      FunType (elab_ty ty1, elab_ty ty2)
+  | Placeholder ->
+      fresh_meta ty.loc `Placeholder
+
+(** In this algorithm type checking is mainly unidirectional, relying on the
+    [infer] funtion, but a {!check} function is provided for convenience.
+*)
+
+(** Elaborate a surface term into a core term, given an expected type. *)
+let rec elab_check (context : context) (tm : tm) (ty : Core.ty) : Core.tm =
+  let tm', ty' = elab_infer context tm in
+  unify tm.loc ty ty';
+  tm'
+
+(** Elaborate a surface term into a core term, inferring its type. *)
+and elab_infer (context : context) (tm : tm) : Core.tm * Core.ty =
+  match tm.data with
+  | Name name -> begin
+      match lookup context name with
+      | Some (index, ty) -> Var index, ty
+      | None -> error tm.loc (Format.asprintf "unbound name `%s`" name)
+  end
+  | Let (def_name, params, def_body_ty, def_body, body) ->
+      let def, def_ty = elab_infer_fun_lit context params def_body_ty def_body in
+      let body, body_ty = elab_infer ((def_name.data, def_ty) :: context) body in
+      Let (def_name.data, def_ty, def, body), body_ty
+  | Ann (tm, ty) ->
+      let ty = elab_ty ty in
+      elab_check context tm ty, ty
+  | BoolLit b -> BoolLit b, BoolType
+  | IfThenElse (head, tm0, tm1) ->
+      let head = elab_check context head BoolType in
+      let ty = fresh_meta tm.loc `IfBranches in
+      let tm0 = elab_check context tm0 ty in
+      let tm1 = elab_check context tm1 ty in
+      BoolElim (head, tm0, tm1), ty
+  | IntLit i -> IntLit i, IntType
+  | FunLit (params, body) ->
+      elab_infer_fun_lit context params None body
+  | FunApp (head, arg) ->
+      let head_loc = head.loc in
+      let head, head_ty = elab_infer context head in
+      let param_ty, body_ty =
+        match head_ty with
+        | FunType (param_ty, body_ty) -> param_ty, body_ty
+        | head_ty ->
+            let param_ty = fresh_meta head_loc `FunParam in
+            let body_ty = fresh_meta head_loc `FunBody in
+            unify head_loc head_ty (FunType (param_ty, body_ty));
+            param_ty, body_ty
+      in
+      let arg = elab_check context arg param_ty in
+      FunApp (head, arg), body_ty
+  | Op2 ((`Eq) as prim, tm0, tm1) ->
+      let tm0 = elab_check context tm0 IntType in
+      let tm1 = elab_check context tm1 IntType in
+      PrimApp (prim, [tm0; tm1]), BoolType
+  | Op2 ((`Add | `Sub | `Mul) as prim, tm0, tm1) ->
+      let tm0 = elab_check context tm0 IntType in
+      let tm1 = elab_check context tm1 IntType in
+      PrimApp (prim, [tm0; tm1]), IntType
+  | Op1 ((`Neg) as prim, tm) ->
+      let tm = elab_check context tm IntType in
+      PrimApp (prim, [tm]), IntType
+  | LetRec (defs, body) ->
+      let context' = ListLabels.map defs ~f:(fun(name, params, ty, _) -> (name.data, elab_infer_fun_signature context name.loc params ty)) in
+      let context = List.append context context' in
```

Comment:
Any reason why we append the recursive bindings at the start of the context as opposed to the end? 🤔

Reply:
Because I'm a dummy who forgot that ML lists are the opposite order to Rust vectors